src/cluster.c

```diff
@@ -251,14 +253,28 @@ void restoreCommand(client *c) {
     }
 
     /* Verify RDB version and data checksum. */
-    if (verifyDumpPayload(c->argv[3]->ptr, sdslen(c->argv[3]->ptr), NULL) == C_ERR) {
+    if (verifyDumpPayload(c->argv[3]->ptr, sdslen(c->argv[3]->ptr), &rdbver) == C_ERR) {
         addReplyError(c, "DUMP payload version or checksum are wrong");
         return;
     }
 
     rioInitWithBuffer(&payload, c->argv[3]->ptr);
-    if (((type = rdbLoadObjectType(&payload)) == -1) ||
-        ((obj = rdbLoadObject(type, &payload, key->ptr, c->db->id, NULL)) == NULL)) {
+    type = rdbLoadObjectType(&payload);
+    if (type == -1) {
+        addReplyError(c, "Bad data format");
+        return;
+    }
+
+    /* If it's a foreign RDB format. Only accept old data types that we know the
+     * meaning of. */
+    if (rdbver >= RDB_FOREIGN_VERSION_MIN && rdbver <= RDB_FOREIGN_VERSION_MAX &&
+        type >= RDB_FOREIGN_TYPE_MIN) {
+        addReplyError(c, "Unsupported foreign data format");
```

Comment:
Would it be more helpful for administrator to figure which data type crept in ?
```suggestion
        addReplyErrorFormat(c, "Unsupported foreign data type: %d", type);
```

Reply:
Sure. It can be a type or opcode.